ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,899 @@
+   */
+  private void checkSanity(CompactionDetails details) {
+    List<String> segmentsUnderCompaction = details.getLogSegmentsUnderCompaction();
+    LogSegment lastSegment = null;
+    // all segments should be available
+    for (String segmentName : segmentsUnderCompaction) {
+      lastSegment = srcLog.getSegment(segmentName);
+      if (lastSegment == null) {
+        throw new IllegalArgumentException(segmentName + " does not exist in the log");
+      }
+    }
+    // last offset to compact should be outside the range of the journal
+    Offset lastSegmentEndOffset = new Offset(lastSegment.getName(), lastSegment.getEndOffset());
+    if (lastSegmentEndOffset.compareTo(srcIndex.journal.getFirstOffset()) >= 0) {
+      throw new IllegalArgumentException("Some of the offsets provided for compaction are within the journal");
+    }
+  }
+
+  /**
+   * Copies valid data from the source log segments into the swap spaces.
+   * <p/>
+   * Splits the compaction into two cycles if there aren't enough swap spaces and completes the copy for the current
+   * cycle.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void copy() throws InterruptedException, IOException, StoreException {
+    setupState();
+    List<String> logSegmentsUnderCompaction = compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction();
+    Offset startOffsetOfLastIndexSegmentForDeleteCheck = getStartOffsetOfLastIndexSegmentForDeleteCheck();
+    FileSpan duplicateSearchSpan = null;
+    if (compactionLog.getCurrentIdx() > 0) {
+      // only records in the the very first log segment in the cycle could have been copied in a previous cycle
+      LogSegment firstLogSegment = srcLog.getSegment(logSegmentsUnderCompaction.get(0));
+      LogSegment prevSegment = srcLog.getPrevSegment(firstLogSegment);
+      if (prevSegment != null) {
+        // duplicate data, if it exists, can only be in the log segment just before the first log segment in the cycle.
+        Offset startOffset = new Offset(prevSegment.getName(), prevSegment.getStartOffset());
+        Offset endOffset = new Offset(prevSegment.getName(), prevSegment.getEndOffset());
+        duplicateSearchSpan = new FileSpan(startOffset, endOffset);
+      }
+    }
+
+    for (String logSegmentName : logSegmentsUnderCompaction) {
+      LogSegment srcLogSegment = srcLog.getSegment(logSegmentName);
+      Offset logSegmentEndOffset = new Offset(srcLogSegment.getName(), srcLogSegment.getEndOffset());
+      if (needsCopying(logSegmentEndOffset) && !copyDataByLogSegment(srcLogSegment, duplicateSearchSpan,
+          startOffsetOfLastIndexSegmentForDeleteCheck)) {
+        if (isActive) {
+          // split the cycle only if there is no shutdown in progress
+          compactionLog.splitCurrentCycle(logSegmentName);
```

Comment:
The function `copyDataByLogSegment()` returns `false` in two cases (both are when the copying is not complete)
1. If the whole log segment could not be copied because there was no more space in the swap spaces.
2. If a shutdown is in progress.

In case 1, the cycle has to be split in order to take compaction to completion. In case 2, there is no splitting required.